### PR TITLE
feat(notifications): implement welcome toast notifications

### PR DESCRIPTION
**Pull request by pbortnick**
## 🔗 Relevant links

- [Preview link](https://dev-portal-git-pbimplement-notifications-hashicorp.vercel.app) 🔎
- [Asana task](https://app.asana.com/0/1202097197789424/1202939992537941/f) 🎟️

## 🗒️ What

Implement welcome notifications on all pages except the homepage.

## 🛠️ How

- Create `makeWelcomeToast` function and run in `useEffect` in _app.tsx
- Steps in function:
1. Return if toast has been permanently dismissed or if session is active
2. Check if welcome toast was ever shown AND if the cookie for it has expired. If so, permanently dismiss the notification.
3. If cookies have never been set, set a cookie that expires in 3 months and another that expires later, which demonstrates that the notification has been viewed.
4. Activate session cookie
5. Display `toast`
- For permanent dismissal send `permanentlyDismiss` as `onDismissCallback` that sets dismissal cookie and removes all other cookies.

## 🧪 Testing

- [ ] Check that toast does not appear on [homepage](https://dev-portal-git-pbimplement-notifications-hashicorp.vercel.app)
- [ ] Navigate to another page and check that toast appears
- [ ] Stay on same page for 15 seconds and check that toast auto-dismisses
- [ ] Refresh page and check that toast does not appear

To check that dismissal works properly: 
- [ ] Comment out `inSession` check & run locally
- [ ] Click on the X in the toast
- [ ] Refresh page and make sure toast does not appear

### CODE REVIEW

File: src/lib/make-welcome-notification.ts
Comment (kendallstrautman) [inline src/lib/make-welcome-notification.ts:12]: 💭 if we want this to be 1 year from GA, perhaps we pass a date object instead?

**Review by kendallstrautman — COMMENTED** [review]
Did some testing and it works as expected! 👏 Just had one note on whether we want to set a hard date for this functionality. I could see it as something we forget to cleanup. 

Do we know if product wants us to release this notification immediately or after the products are launched?

File: src/lib/make-welcome-notification.ts
Comment (kendallstrautman) [inline src/lib/make-welcome-notification.ts:43]: Could be easier to read if we store some of these numbers in variables. Esp the 15000 one.

File: src/lib/make-welcome-notification.ts
Comment (pbortnick) [reply src/lib/make-welcome-notification.ts]: Yeah sounds good! We can also add a check to see if the current date is later than the date we want to remove these on. You can see that [here](https://github.com/hashicorp/dev-portal/pull/995/commits/95791b152223a33dacd60f704d0c168caf0b3a93#diff-b043befe629def842d164cb756567813f1e0dcacc5ea16d005ae2a99d8b7188dR38). I am confirming now that we won't want these in ~a year and the other question you asked regarding release 👍

File: src/components/opt-in-out/helpers/make-beta-welcome-toast.tsx
Comment (kendallstrautman) [inline src/components/opt-in-out/helpers/make-beta-welcome-toast.tsx:12]: 🤔 are they really sure they want to remove this toast while we still have beta opt in flows in effect? Seems odd to remove.

Comment (kendallstrautman) [reply src/components/opt-in-out/helpers/make-beta-welcome-toast.tsx:12]: I won't fight it but it just seems weird to me lol

File: src/lib/make-welcome-notification.ts
Comment (kendallstrautman) [inline src/lib/make-welcome-notification.ts:9]: nice 😄

File: src/lib/make-welcome-notification.ts
Comment (kendallstrautman) [inline src/lib/make-welcome-notification.ts:27]:
```suggestion
	const hasDismissedWelcomeToast = Cookies.get(DISMISS_COOKIE)
```
naming nit

File: src/lib/make-welcome-notification.ts
Comment (kendallstrautman) [inline src/lib/make-welcome-notification.ts:39]: 🤔 we aren't currently doing anything with this welcomeToastCookie. Do we still need it if we just always want to render one toast until the session cookie is set?

File: src/lib/make-welcome-notification.ts
Comment (kendallstrautman) [inline src/lib/make-welcome-notification.ts:50]: Interestingly, if you navigate to another page while the toast is displayed, the `onDismissCallback` fires and sets the perma dismiss cookie. This might not be an issue, but wanted to note it.

Comment (pbortnick) [reply src/lib/make-welcome-notification.ts:50]: Interesting. I do think that is an issue

Comment (pbortnick) [reply src/lib/make-welcome-notification.ts:50]: I think this can be resolved by adding `dismissOnRouteChange: false` to the toast options 👍 . Good catch!

File: src/components/opt-in-out/helpers/make-beta-welcome-toast.tsx
Comment (pbortnick) [reply src/components/opt-in-out/helpers/make-beta-welcome-toast.tsx:12]: 🤷‍♀️

File: src/lib/make-welcome-notification.ts
Comment (pbortnick) [reply src/lib/make-welcome-notification.ts]: Yeah you're right we don't need it if we are adding the date check. Alternatively we can keep it with the proper expiration, remove the date check, and then check if the `welcomeToastCookie` exists before rendering. What do you think?
https://github.com/hashicorp/dev-portal/pull/995/commits/ef580a5838b31ec49a7645c21f190ca0dd1f8dd4#diff-b043befe629def842d164cb756567813f1e0dcacc5ea16d005ae2a99d8b7188dR42

Comment (kendallstrautman) [reply src/lib/make-welcome-notification.ts]: I think retaining the date check and just removing this cookie would be reasonable! That limits the amount of cookies we're storing 😄

**Review by kendallstrautman — APPROVED** [review]
Approving to unblock! But we should address that small welcome cookie discussion before merging 💯 Great work

File: src/lib/make-welcome-notification.ts
Comment (kendallstrautman) [reply src/lib/make-welcome-notification.ts]: 🕵️ 👏

File: src/lib/make-welcome-notification.ts
Comment (pbortnick) [reply src/lib/make-welcome-notification.ts]: I'm fine with that too since we are still involving 2 other cookies in the function 👍

File: src/lib/make-welcome-notification.ts
Comment (pbortnick) [inline src/lib/make-welcome-notification.ts:19]: @kendallstrautman changed it back here 📆 😸

**Pull request merged**
